flink-tests/src/test/scala/org/apache/flink/api/scala/migration/StatefulJobSavepointMigrationITCase.scala

```diff
@@ -38,78 +38,95 @@ import org.apache.flink.streaming.api.functions.source.SourceFunction
+import org.apache.flink.test.util.MigrationTest
+import org.apache.flink.test.util.MigrationTest.ParameterizedSnapshotsGenerator
 import org.apache.flink.util.Collector
 
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.junit.runners.Parameterized
 
+import javax.annotation.Nullable
+
 import java.util
-import java.util.stream.Collectors
+import java.util.Collections
+import java.util.function.BiFunction
 
 import scala.util.{Failure, Try}
 
 object StatefulJobSavepointMigrationITCase {
 
-  // TODO increase this to newer version to create and test snapshot migration for newer versions
-  val currentVersion = FlinkVersion.v1_16
-
-  // TODO change this to CREATE_SNAPSHOT to (re)create binary snapshots
-  // TODO Note: You should generate the snapshot based on the release branch instead of the
-  // master.
-  val executionMode = ExecutionMode.VERIFY_SNAPSHOT
-
   @Parameterized.Parameters(name = "Test snapshot: {0}")
-  def parameters: util.Collection[SnapshotSpec] = {
+  def createSpecsForTestRuns: util.Collection[SnapshotSpec] =
+    internalParameters(null)
+
+  def createSpecsForTestDataGeneration(version: FlinkVersion): util.Collection[SnapshotSpec] =
+    internalParameters(version)
+
+  private def internalParameters(
+      @Nullable targetGeneratingVersion: FlinkVersion): util.Collection[SnapshotSpec] = {
+    val getFlinkVersions =
+      new BiFunction[FlinkVersion, FlinkVersion, util.Collection[FlinkVersion]] {
+        override def apply(
+            minInclVersion: FlinkVersion,
+            maxInclVersion: FlinkVersion): util.Collection[FlinkVersion] = if (
+          targetGeneratingVersion != null
+        )
+          Collections.singleton(targetGeneratingVersion)
+        else
+          FlinkVersion.rangeOf(minInclVersion, maxInclVersion)
+      }
+
     // Note: It is not safe to restore savepoints created in a Scala applications with Flink
     // version 1.7 or below. The reason is that up to version 1.7 the underlying Scala serializer
     // used names of anonymous classes that depend on the relative position/order in code, e.g.,
     // if two anonymous classes, instantiated inside the same class and from the same base class,
     // change order in the code their names are switched.
     // As a consequence, changes in code may result in restore failures.
     // This was fixed in version 1.8, see: https://issues.apache.org/jira/browse/FLINK-10493
-    var parameters: util.List[SnapshotSpec] = new util.LinkedList[SnapshotSpec]()
+    val parameters: util.List[SnapshotSpec] = new util.LinkedList[SnapshotSpec]()
     parameters.addAll(
       SnapshotSpec.withVersions(
         StateBackendLoader.MEMORY_STATE_BACKEND_NAME,
         SnapshotType.SAVEPOINT_CANONICAL,
-        FlinkVersion.rangeOf(FlinkVersion.v1_8, FlinkVersion.v1_13)))
+        getFlinkVersions.apply(FlinkVersion.v1_8, FlinkVersion.v1_13)
```

Comment:
Here as well: The old code filters for the most-recently published version. The new code doesn't do that anymore.